src/fabric/src/fabric2_fdb.erl

```diff
@@ -346,12 +351,97 @@ delete(#{} = Db) ->
     } = ensure_current(Db),
 
     DbKey = erlfdb_tuple:pack({?ALL_DBS, DbName}, LayerPrefix),
-    erlfdb:clear(Tx, DbKey),
-    erlfdb:clear_range_startswith(Tx, DbPrefix),
+    DoRecovery = fabric2_util:do_recovery(),
```

Comment:
I think splitting this up into three helpers would be the cleanest API. Something like this:

```erlang
delete(Db, Options) ->
    DoRecovery = fabric2_util:do_recovery(),
    DeletedAt = fabric2_util:get_value(deleted_at, Options),

    case {DoRecovery, DeletedAt} of
        {true, undefined} -> soft_delete_db(Db);
        {false, undefined} -> hard_delete_db(Db);
        {_, TimeStamp} -> remove_deleted_db(Db, Timestamp)
    end.


soft_delete_db(Db) ->
    #{
        name := DbName,
        tx := Tx,
        layer_prefix := LayerPrefix
    } = ensure_current(Db),

    DbKey = erlfdb_tuple:pack({?ALL_DBS, DbName}, LayerPrefix)
    Timestamp = list_to_binary(fabric2_util:iso8601_timestamp()),
    DeletedDbKeyTuple = {?DELETED_DBS, DbName, Timestamp},
    DeletedDbKey = erlfdb_tuple:pack(DeletedDbKeyTuple, LayerPrefix),
    case erlfdb:wait(erlfdb:get(Tx, DeletedDbKey)) of
        not_found ->
            erlfdb:set(Tx, DeletedDbKey, DbPrefix),
            erlfdb:clear(Tx, DbKey);
        _Val ->
            erlang:error({deleted_database_exists, DbName})
    end,
    bump_metadata_version(Tx),
    ok.


hard_delete_db(Db) ->
    #{
        name := DbName,
        tx := Tx,
        layer_prefix := LayerPrefix,
        db_prefix := DbPrefix
    } = ensure_current(Db),

    DbKey = erlfdb_tuple:pack({?ALL_DBS, DbName}, LayerPrefix)

    erlfdb:clear(Tx, DbKey),
    erlfdb:clear_range_startswith(Tx, DbPrefix),
    bump_metadata_version(Tx),
    ok.


remove_deleted_db(Db, Timestamp) ->
    #{
        name := DbName,
        tx := Tx,
        layer_prefix := LayerPrefix
    } = ensure_current(Db0, false),

    DeletedDbTupleKey = {?DELETED_DBS, DbName, TimeStamp},
    DeletedDbKey =  erlfdb_tuple:pack(DeletedDbTupleKey, LayerPrefix),
    case erlfdb:wait(erlfdb:get(Tx, DeletedDbKey)) of
        not_found ->
            erlang:error({not_found, missing_timestamp});
        DbPrefix ->
            erlfdb:clear(Tx, DeleteDbKey),
            erlfdb:clear_range_startswith(Tx, DbPrefix)
    end,
    ok.
```

Though the three helpers get moved down into the helper function section.

Reply:
I split `delete` into `soft_delete_db/1` and `hard_delete_db/1`. For `remove_deleted_db`, it is a separate call from `fabric2_db:delete/2`.

```
delete(#{} = Db) ->
    DoRecovery = fabric2_util:do_recovery(),
    case DoRecovery of
        true -> soft_delete_db(Db);
        false -> hard_delete_db(Db)
    end.
```